### PR TITLE
Manage Appnet versioning

### PR DESCRIPTION
**Pull request by mythri-garaga**
### Summary

This PR includes changes to handle the major version updates to AppNet image version.

### Implementation details

- ` agent/engine/serviceconnect/version_linux.go`:
	New version_linux file contains supportedAppnetInterfaceVerToCapability map to track all the AppNet interface versions supported by ECS Agent. Added getSupportedAppnetInterfaceVersions() function that returns the supported AppNet interface versions by ECS Agent from highest to lowest version.
- `agent/engine/serviceconnect/manager_linux.go`:
         - Updated LoadImage() to iterate through supported Appnet interface versions list from highest to lowest order and check if the corresponding tar file exists before loading, if loading is successfully `AppnetInterfaceVersion` for the runtime.
        - Added a new function GetSupportedAppnetInterfaceVerToCapabilities() to return service connect capabilities supported by ECS Agent to register for a selected AppNet version.
- Added unit tests to cover the above functions and updated existing texts to handle mock calls to GetSupportedAppnetInterfaceVerToCapabilities(). Will add a test to check if right AppNet version is being loaded in the upcoming CR which includes changes to restart relay on AppNet image update.

### Testing

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/serviceconnect/version_linux.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// getSupportedAppnetInterfaceVersion returns the latest supported AppNet interface version
+// by looking for the AppNet agent tar name available on the instance
+// from supportedAppnetInterfaceVerToCapability map from highest to lowest version
+func getSupportedAppnetInterfaceVersion() string {
+	supportedAppnetInterfaceVersions := make([]string, 0, len(supportedAppnetInterfaceVerToCapability))
+	for supportedAppnetInterfaceVersion := range supportedAppnetInterfaceVerToCapability {
+		supportedAppnetInterfaceVersions = append(supportedAppnetInterfaceVersions, supportedAppnetInterfaceVersion)
+	}
+	sort.Sort(sort.Reverse(sort.StringSlice(supportedAppnetInterfaceVersions)))
+
```
Comment (yinyic) [inline agent/engine/serviceconnect/version_linux.go:44]: I think it would be better if we return the list right here, and let the loader go through the list and validate the existence. So something like
```
for each appnetInterfaceVersion in getSupportedAppnetInterfaceVersions():  // from highest to lowest
  if tar file does not exist, continue
  if loading tar file failed, continue
  # if we reach here we've successfully loaded a tar file
  break
```
That's because
- it will be a cleaner separation between static capability and runtime validation (which needs to be performed by loader anyway)
- there might be cases where the tar file exists but is corrupted, and in such case we should let loader keeping trying older version of AppNet

File: agent/app/agent_capability.go [hunk 2]
```diff
@@ -458,7 +458,11 @@ func (agent *ecsAgent) appendServiceConnectCapabilities(capabilities []*ecs.Attr
 		}
 	}
```
Comment (yinyic) [inline agent/app/agent_capability.go:460]: More details in the comment for `getSupportedAppnetInterfaceVersion` - I think we should let loader go through the list of supported appnet versions and keep trying until it successfully loads a tar, and it returns that version.

File: agent/app/agent_capability.go [hunk 3]
```diff
@@ -458,7 +458,11 @@ func (agent *ecsAgent) appendServiceConnectCapabilities(capabilities []*ecs.Attr
...
 
-	return appendNameOnlyAttribute(capabilities, attributePrefix+capabilityServiceConnect)
+	supportedAppnetInterfaceVerToCapabilities := agent.serviceconnectManager.GetSupportedAppnetInterfaceVerToCapabilities()
```
Comment (yinyic) [inline agent/app/agent_capability.go:461]: Ditto - the knowledge of "which appnet version are we actually using" should come from loader as it's a runtime decision.

File: agent/engine/serviceconnect/manager_linux.go [hunk 4]
```diff
@@ -139,9 +143,11 @@ func NewManager() Manager {
...
 		logPathECSAgentRoot: fmt.Sprintf(defaultECSAgentLogPathForSC, getECSAgentLogPathContainer()),
 
-		AgentContainerImageName:   defaultAgentContainerImageName,
-		AgentContainerTag:         defaultAgentContainerTag,
-		AgentContainerTarballPath: defaultAgentContainerTarballPath,
+		AgentContainerImageName: defaultAgentContainerImageName,
+		AgentContainerTag:       defaultAgentContainerTag,
+		AgentContainerVersion:   getSupportedAppnetInterfaceVersion(),
+		AgentContainerTarballPath: fmt.Sprintf(defaultAgentContainerTarballPathFormat,
+			getSupportedAppnetInterfaceVersion()),
```
Comment (yinyic) [inline agent/engine/serviceconnect/manager_linux.go:150]: Ditto - the final appnet agent version being loaded should come from loader.

File: agent/engine/serviceconnect/manager_linux.go [hunk 5]
```diff
@@ -372,17 +370,30 @@ func (m *manager) AugmentInstanceContainer(task *apitask.Task, container *apicon
...
 func (agent *manager) LoadImage(ctx context.Context, _ *config.Config, dockerClient dockerapi.DockerClient) (*types.ImageInspect, error) {
-	logger.Debug("Loading appnet agent container tarball:", logger.Fields{
-		field.Image: agent.AgentContainerTarballPath,
-	})
-	if err := loader.LoadFromFile(ctx, agent.AgentContainerTarballPath, dockerClient); err != nil {
-		return nil, err
+	var err error
+	for _, supportedAppnetInterfaceVersion := range getSupportedAppnetInterfaceVersions() {
+		agentContainerTarballPath := fmt.Sprintf(defaultAgentContainerTarballPathFormat, supportedAppnetInterfaceVersion)
+		if _, err = os.Stat(agentContainerTarballPath); err == nil {
```
Comment (yinyic) [inline agent/engine/serviceconnect/manager_linux.go:380]: Maybe also check for non-nil error here to be more consistent with load error below?
```
if _, err = os.Stat(agentContainerTarballPath); err != nil {
 // log a warning message
 continue
}
if loadErr...
```

File: agent/engine/serviceconnect/manager_linux.go [hunk 6]
```diff
@@ -372,17 +370,30 @@ func (m *manager) AugmentInstanceContainer(task *apitask.Task, container *apicon
...
+			if loadErr := loader.LoadFromFile(ctx, agentContainerTarballPath, dockerClient); err != nil {
+				logger.Warn("Unable to load appnet agent container tarball:", logger.Fields{
+					field.Image: agentContainerTarballPath,
+					field.Error: loadErr,
+				})
+				continue
+			}
+			agent.AppnetInterfaceVersion = supportedAppnetInterfaceVersion
+			imageName, _ := agent.GetLoadedImageName()
+			return loader.GetContainerImage(imageName, dockerClient)
```
Comment (yinyic) [inline agent/engine/serviceconnect/manager_linux.go:393]: Let's log an info message here to state the tar we've successfully loaded, and what we are assuming the loaded image name to be

File: agent/engine/serviceconnect/manager_linux.go [hunk 7]
```diff
@@ -402,3 +414,16 @@ func getECSAgentLogPathContainer() string {
...
+// Suppose we decide to register ecs.service-connect.v2 capability for new AppNet version (ex: 1.24.0.0),
+// now if ecs.service-connect.*v1* is continuously being supported by 1.24.0.0,
+// we will then register multiple capabilities.
+// {
+//    "v1": ["ecs.service-connect.v1"],
+//    "v2": ["ecs.service-connect.v1", "ecs.service-connect.v2"]
+// }
+func (agent *manager) GetSupportedAppnetInterfaceVerToCapabilities() []string {
+	return supportedAppnetInterfaceVerToCapability[agent.AppnetInterfaceVersion]
+}
```
Comment (yinyic) [inline agent/engine/serviceconnect/manager_linux.go:429]: Nit - maybe rename to `GetCapabilitiesForAppnetInterfaceVersion`

File: agent/engine/serviceconnect/manager_linux.go [hunk 8]
```diff
@@ -402,3 +414,16 @@ func getECSAgentLogPathContainer() string {
...
+// supported by ECS Agent to register for a selected AppNet version.
+// Suppose we decide to register ecs.service-connect.v2 capability for new AppNet version (ex: 1.24.0.0),
+// now if ecs.service-connect.*v1* is continuously being supported by 1.24.0.0,
+// we will then register multiple capabilities.
+// {
+//    "v1": ["ecs.service-connect.v1"],
+//    "v2": ["ecs.service-connect.v1", "ecs.service-connect.v2"]
+// }
+func (agent *manager) GetSupportedAppnetInterfaceVerToCapabilities() []string {
+	return supportedAppnetInterfaceVerToCapability[agent.AppnetInterfaceVersion]
```
Comment (yinyic) [inline agent/engine/serviceconnect/manager_linux.go:428]: I feel that it may better to have caller (`appendServiceConnectCapabilities`) explicitly pass in the appnet version. Currently it first calls loader, loader will set the global appnet version, which is used by this method, so we are introducing some dependencies between those two methods which couldn't been independent. 

Let's maybe have below in `appendServiceConnectCapabilities` instead:
```
loadedVer, err := agent.LoadImage(...)
// handle err
capabilities := agent.GetCapabilitiesForAppnetInterfaceVersion(loadedVer)
// if nil, throw warning
// append the capabilities
```
Overall feels a bit more clear to me. What do you think?

File: agent/engine/serviceconnect/version_linux.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+// getSupportedAppnetInterfaceVersions returns the all the supported AppNet interface versions
+// by ECS Agent from highest to lowest version
+func getSupportedAppnetInterfaceVersions() []string {
+	supportedAppnetInterfaceVersions := make([]string, 0, len(supportedAppnetInterfaceVerToCapability))
+	for supportedAppnetInterfaceVersion := range supportedAppnetInterfaceVerToCapability {
+		supportedAppnetInterfaceVersions = append(supportedAppnetInterfaceVersions, supportedAppnetInterfaceVersion)
+	}
+	sort.Sort(sort.Reverse(sort.StringSlice(supportedAppnetInterfaceVersions)))
```
Comment (yinyic) [inline agent/engine/serviceconnect/version_linux.go:40]: Will this become problematic if we are comparing `v10` to `v2`?

File: agent/engine/serviceconnect/manager_linux.go (hunk 5 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go:380]: Will add a check and a warning log here, thanks!

File: agent/engine/serviceconnect/version_linux.go (hunk 9 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/version_linux.go]: Good catch! I have updated the sorting logic to handle it correctly.
Thank you!

File: agent/engine/serviceconnect/manager_linux.go (hunk 7 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go]: Done:)

File: agent/engine/serviceconnect/manager_linux.go (hunk 6 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go]: Added the info message.

File: agent/engine/serviceconnect/manager_linux.go (hunk 8 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go]: The LoadImage() function is used to load pause container image as well, thought returning version might not add much value in that context. 
I have added a get and set functions to manage the appnet version, hope it makes it a bit clear, please let me know:)

File: agent/logger/field/constants.go [hunk 10]
```diff
@@ -36,4 +36,5 @@ const (
 	Volume        = "volume"
 	Time          = "time"
 	NetworkMode   = "networkMode"
+	Tar           = "tar"
```
Comment (yinyic) [inline agent/logger/field/constants.go:39]: I'm not sure if we will be referencing AppNet tar frequently enough log to merit adding it as a logger field, seems like only loader uses it... And also because the fields so far have been general task properties, but appnet tar name not so much. 
I think I'd probably rather just use formatted string in the few places we want to log the tar file name.

File: agent/engine/serviceconnect/manager_linux.go [hunk 11]
```diff
@@ -372,17 +370,43 @@ func (m *manager) AugmentInstanceContainer(task *apitask.Task, container *apicon
...
-		field.Image: agent.AgentContainerTarballPath,
-	})
-	if err := loader.LoadFromFile(ctx, agent.AgentContainerTarballPath, dockerClient); err != nil {
-		return nil, err
+	var err error
+	for _, supportedAppnetInterfaceVersion := range getSupportedAppnetInterfaceVersions() {
+		agentContainerTarballPath := fmt.Sprintf(defaultAgentContainerTarballPathFormat, supportedAppnetInterfaceVersion)
+		if _, err = os.Stat(agentContainerTarballPath); err != nil {
+			logger.Warn("Agent container tarball unavailable:", logger.Fields{
+				field.Tar:   agentContainerTarballPath,
```
Comment (yinyic) [inline agent/engine/serviceconnect/manager_linux.go:386]: Ditto - Not sure we want to add it to logger field (and the ones below)

File: agent/engine/serviceconnect/manager_linux.go [hunk 12]
```diff
@@ -78,9 +78,9 @@ const (
 	defaultAdminStatsRequest = httpRequestPrefix + "/stats/prometheus?usedonly&filter=metrics_extension&delta"
 	defaultAdminDrainRequest = httpRequestPrefix + "/drain_listeners?inboundonly"
 
-	defaultAgentContainerTarballPath = "/managed-agents/serviceconnect/appnet_agent.interface-v1.tar"
-	defaultAgentContainerImageName   = "appnet_agent"
-	defaultAgentContainerTag         = "service_connect.v1"
+	defaultAgentContainerImageName         = "ecs-service-connect-agent"
+	defaultAgentContainerTagFormat         = "interface-%s"
+	defaultAgentContainerTarballPathFormat = "ecs-service-connect-agent.interface-%s.tar"
```
Comment (yinyic) [inline agent/engine/serviceconnect/manager_linux.go:83]: Sorry if my previous comment didn't make it clear - we still need the `/managed-agents/serviceconnect/` path, just that the tar file name is now `ecs-service-connect-agent.interface-%s.tar` instead of `appnet_agent.interface-%s.tar`

File: agent/engine/serviceconnect/version_linux.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+		func(i, j int) bool {
+			return sortVersion(supportedAppnetInterfaceVersions[i]) > sortVersion(supportedAppnetInterfaceVersions[j])
+		},
+	)
+	return supportedAppnetInterfaceVersions
+}
+
+// sortVersion returns a version sort key in numeric order.
+// For example, v20, v11, v10, v2, v1,...
+func sortVersion(version string) int {
```
Comment (yinyic) [inline agent/engine/serviceconnect/version_linux.go:68]: I'd probably rename it - this function itself simply parses the int value from version string, and doesn't do any sorting on its own. Maybe `getVersionNumber` or something more appropriate.

File: agent/engine/serviceconnect/version_linux.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	for supportedAppnetInterfaceVersion := range supportedAppnetInterfaceVerToCapability {
+		allowedVersionPattern, err := regexp.Compile(appNetInterfaceVersionPattern)
+		if err != nil {
+			logger.Warn("Unable to compile regex pattern:", logger.Fields{
+				field.Error: err,
+			})
+		}
+		if !allowedVersionPattern.MatchString(supportedAppnetInterfaceVersion) {
+			logger.Warn("Appnet interface version does not match the expected pattern, ex: `v1`")
+		}
```
Comment (yinyic) [inline agent/engine/serviceconnect/version_linux.go:54]: I feel like this should be something caught in a unit test instead of runtime. This is all static information known at compile time.

File: agent/engine/serviceconnect/manager_linux.go (hunk 12 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go:83]: my bad, I updated it incorrectly by mistake. Will fix it

File: agent/engine/serviceconnect/version_linux.go (hunk 14 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/version_linux.go]: Agreed! Moved the validation to a unit test.

**Review by yinyic — APPROVED** [review]
LGTM! Would like to see we verify the integration with appnet rpm with a manual test, mainly to check if all the tarball path, image name and tags are as expected.

**Comment by mythri-garaga** [comment]
> LGTM! Would like to see we verify the integration with appnet rpm with a manual test, mainly to check if all the tarball path, image name and tags are as expected.

Tested these changes by building a test AMI. ECS Agent was able to load the new appnet image as expect - 
```
level=debug time=2022-10-10T19:25:18Z msg="Inspecting container image: " amazon/amazon-ecs-pause:0.1.0="amazon/amazon-ecs-pause:0.1.0"
level=debug time=2022-10-10T19:25:18Z msg="Inspecting container image: " ecs-service-connect-agent:interface-="ecs-service-connect-agent:interface-"
level=debug time=2022-10-10T19:25:18Z msg="Loading appnet agent container tarball: /managed-agents/serviceconnect/ecs-service-connect-agent.interface-v1.tar"
level=info time=2022-10-10T19:25:25Z msg="Successfully loaded appnet agent container tarball: /managed-agents/serviceconnect/ecs-service-connect-agent.interface-v1.tar" image="ecs-service-connect-agent:interface-v1"
level=debug time=2022-10-10T19:25:25Z msg="Inspecting container image: " ecs-service-connect-agent:interface-v1="ecs-service-connect-agent:interface-v1"
```

File: agent/engine/serviceconnect/manager_linux.go [hunk 15]
```diff
@@ -115,9 +115,9 @@ type manager struct {
 	// Http path + params to make a drain request of AppNetAgent
 	adminDrainRequest string
 
-	AgentContainerImageName   string
-	AgentContainerTag         string
-	AgentContainerTarballPath string
+	AgentContainerImageName string
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/manager_linux.go:118]: Q. are these variables needed to be exported? should we change them to lower-case?

File: agent/engine/serviceconnect/manager_linux.go [hunk 16]
```diff
@@ -372,17 +370,40 @@ func (m *manager) AugmentInstanceContainer(task *apitask.Task, container *apicon
...
-	logger.Debug("Loading appnet agent container tarball:", logger.Fields{
-		field.Image: agent.AgentContainerTarballPath,
-	})
-	if err := loader.LoadFromFile(ctx, agent.AgentContainerTarballPath, dockerClient); err != nil {
-		return nil, err
+	var err error
+	for _, supportedAppnetInterfaceVersion := range getSupportedAppnetInterfaceVersions() {
+		agentContainerTarballPath := fmt.Sprintf(defaultAgentContainerTarballPathFormat, supportedAppnetInterfaceVersion)
+		if _, err = os.Stat(agentContainerTarballPath); err != nil {
+			logger.Warn(fmt.Sprintf("Agent container tarball unavailable: %s", agentContainerTarballPath), logger.Fields{
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/manager_linux.go:385]: Not blocking. Could we add "Appnet" or "AppNet" before "Agent container tarball unavailable" to avoid mixing it up with ECS Agent or other Agents? And could we have a standard for using "Appnet", "appnet" or "AppNet" in all warning/error messages?  Thanks!

File: agent/engine/serviceconnect/manager_linux.go [hunk 17]
```diff
@@ -391,9 +412,14 @@ func (agent *manager) IsLoaded(dockerClient dockerapi.DockerClient) (bool, error
 }
 
 func (agent *manager) GetLoadedImageName() (string, error) {
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/manager_linux.go:414]: Should we remove returning `error` from the return value since it will always be nil?

File: agent/app/agent_capability.go [hunk 18]
```diff
@@ -457,8 +457,18 @@ func (agent *ecsAgent) appendServiceConnectCapabilities(capabilities []*ecs.Attr
 			return capabilities
 		}
 	}
-
-	return appendNameOnlyAttribute(capabilities, attributePrefix+capabilityServiceConnect)
+	loadedVer := agent.serviceconnectManager.GetLoadedAppnetVersion()
```
Comment (chienhanlin) [inline agent/app/agent_capability.go:460]: Should we add a check here just in case `GetLoadedAppnetVersion()` returns  `nil `or an `empty` string?

File: agent/engine/serviceconnect/manager_other.go [hunk 19]
```diff
@@ -69,3 +69,11 @@ func (*manager) GetLoadedImageName() (string, error) {
 		"appnetAgent container get image name: unsupported platform: %s/%s",
 		runtime.GOOS, runtime.GOARCH))
 }
+
+func (*manager) GetLoadedAppnetVersion() string {
+	return ""
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/manager_other.go:74]: Q. should we also return a `NewUnsupportedPlatformError` here?

File: agent/engine/serviceconnect/version_linux.go [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+		},
+	)
+	return supportedAppnetInterfaceVersions
+}
+
+// getVersionNumber returns a version sort key in numeric order.
+func getVersionNumber(version string) int {
+	versionNumber := strings.TrimPrefix(version, "v")
+	num, err := strconv.Atoi(versionNumber)
+	if err != nil {
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/version_linux.go:58]: Q. should we consider dropping versionNumber that failed to be parsed, and removing it from the `supportedAppnetInterfaceVersions` before returning by `getSupportedAppnetInterfaceVersions()`?

File: agent/engine/serviceconnect/version_linux_test.go [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+
+func TestSupportedAppnetInterfaceVersion(t *testing.T) {
+	allowedVersionPattern, err := regexp.Compile(appNetInterfaceVersionPattern)
+	for supportedAppnetInterfaceVersion := range supportedAppnetInterfaceVerToCapability {
+		require.NoError(t, err, "Error while compiling regex pattern")
+		require.True(t, allowedVersionPattern.MatchString(supportedAppnetInterfaceVersion),
+			"Appnet interface version does not match the expected pattern, ex:`v1`")
+	}
+}
+func TestGetSupportedAppnetInterfaceVersion(t *testing.T) {
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/version_linux_test.go:36]: Could we add negative tests cases to verify how errors will be handled : )

File: agent/engine/serviceconnect/manager_linux_test.go [hunk 22]
```diff
@@ -150,6 +150,28 @@ func TestGetECSAgentLogPathContainer(t *testing.T) {
 			}
 			actualPath := getECSAgentLogPathContainer()
 			assert.Equal(t, tc.expected, actualPath)
+
+		})
+	}
+}
+
+func TestGetSupportedAppnetInterfaceVerToCapabilities(t *testing.T) {
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/manager_linux_test.go:158]: Could we add negative tests cases to verify how errors will be handled : )

File: agent/engine/serviceconnect/manager_linux.go (hunk 16 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go:385]: Ack! 

Since all these callouts are non blocking, will address them in the upcoming PR related to AppNet reload(including all the below comments as well) :)
Thank you!

File: agent/engine/serviceconnect/manager_linux.go (hunk 15 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go:118]: Ya  I can change this to `agentContainerImageName` since it not being exported.

File: agent/app/agent_capability.go (hunk 18 above)
Comment (mythri-garaga) [reply agent/app/agent_capability.go:460]: We have warning log on 463 and 464 which will log `loadedVer` if it's empty(one of the scenario).

File: agent/engine/serviceconnect/manager_linux.go (hunk 17 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go:414]: we return error for unsupported platforms so I think good to keep it

File: agent/engine/serviceconnect/manager_other.go (hunk 19 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_other.go:74]: sure, will make this change!

File: agent/engine/serviceconnect/manager_linux_test.go (hunk 22 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux_test.go:158]: GetSupportedAppnetInterfaceVerToCapabilities() returns errors only for unsupported platforms so adding a test case that handles a empty string for appNetAgentVersion.

File: agent/engine/serviceconnect/version_linux.go (hunk 20 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/version_linux.go:58]: `getVersionNumber()` is only used by `getSupportedAppnetInterfaceVersions()` and there is already a unit test to validate the input for `supportedAppnetInterfaceVersions` so strconv.Atoi(versionNumber) will not return zero in this scenario.

File: agent/engine/serviceconnect/version_linux_test.go (hunk 21 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/version_linux_test.go:36]: same reason as above, getSupportedAppnetInterfaceVersions() returns error when it cannot parse int with strconv.Atoi(versionNumber) but versions are explicitly validated in a unit case so negative test cases are not required here.

File: agent/engine/serviceconnect/manager_linux.go [hunk 23]
```diff
@@ -372,17 +370,40 @@ func (m *manager) AugmentInstanceContainer(task *apitask.Task, container *apicon
...
+		if loadErr = loader.LoadFromFile(ctx, agentContainerTarballPath, dockerClient); loadErr != nil {
+			logger.Warn(fmt.Sprintf("Unable to load appnet agent container tarball: %s", agentContainerTarballPath),
+				logger.Fields{
+					field.Error: loadErr,
+				})
+			continue
+		}
+		agent.setLoadedAppnetVerion(supportedAppnetInterfaceVersion)
+		imageName, _ := agent.GetLoadedImageName()
+		logger.Info(fmt.Sprintf("Successfully loaded appnet agent container tarball: %s", agentContainerTarballPath),
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/manager_linux.go:400]: Can we update the log message to `AppNet agent` in the follow up PR? Thanks!

File: agent/engine/serviceconnect/manager_linux.go [hunk 24]
```diff
@@ -372,17 +370,40 @@ func (m *manager) AugmentInstanceContainer(task *apitask.Task, container *apicon
...
+			logger.Fields{
+				field.Image: imageName,
+			})
+		return loader.GetContainerImage(imageName, dockerClient)
 	}
-
-	imageName, _ := agent.GetLoadedImageName()
-	return loader.GetContainerImage(imageName, dockerClient)
+	return nil, loadErr
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/manager_linux.go:406]: Not blocking and more a question. In addition to show the warning log message, should we aggregate loadErr got from the loop? Only the last error, which might for the lowest AppNet agent version, will be returned with the current logic.

File: agent/engine/serviceconnect/manager_other.go [hunk 25]
```diff
@@ -69,3 +69,15 @@ func (*manager) GetLoadedImageName() (string, error) {
 		"appnetAgent container get image name: unsupported platform: %s/%s",
 		runtime.GOOS, runtime.GOARCH))
 }
+
+func (*manager) GetLoadedAppnetVersion() (string, error) {
+	return "", loader.NewUnsupportedPlatformError(fmt.Errorf(
+		"appnetAgent container get loaded appnet version: unsupported platform: %s/%s",
```
Comment (chienhanlin) [inline agent/engine/serviceconnect/manager_other.go:75]: Can we update the log message to `AppNet agent` for line 75 and line 81 in the follow up PR? Thanks!

**Review by chienhanlin — APPROVED** [review]

File: agent/engine/serviceconnect/manager_other.go (hunk 25 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_other.go:75]: Was following the same convention as other functions in this file...

File: agent/engine/serviceconnect/manager_linux.go (hunk 24 above)
Comment (mythri-garaga) [reply agent/engine/serviceconnect/manager_linux.go:406]: On a second thought, error aggregation is not required since we log the error message for each version we try load.

**Comment by fabrizzioalco** [comment]
Hello! Where does the `appnet_agent.interface-v1.tar` coming from? Im having an issue where installing ecs agentl, the file is not present on the ec2 instance
